File: Sorting/MergeSort.cpp
#include<bits/stdc++.h>
using namespace std;

void merge(vector<int>&arr,int start,int mid,int end){
    vector<int>temp;
    int i = start;
    int j = mid+1;
    while(i<=mid && j<=end){
        if(arr[i]<=arr[j]){
            temp.push_back(arr[i]);
            i++;
        }else{
            temp.push_back(arr[j]);
            j++;
        }
    }
    //for left half if any element still remains
    while(i<=mid){
        temp.push_back(arr[i]);
        i++;
    }
    while(j<=end){
        temp.push_back(arr[j]);
        j++;
    }
  //copying elements from temp to our array
  for(int idx = 0; idx<temp.size();idx++){
    arr[idx+start] = temp[idx];
  }
}

void mergeSort(vector<int>&arr,int start,int end){
    if(start<end){
        int mid = start+(end-start)/2;
        mergeSort(arr,start,mid);
        mergeSort(arr,mid+1,end);
        merge(arr,start,mid,end);
    }
}

int main(){
    vector<int>arr = {76,89,1,4,8,9};
    int n = arr.size()-1;
    mergeSort(arr,0,n);
    for(int it : arr){
        cout<<it<<endl;
    }

}

File: Sorting/MergeSort.py
def merge(arr,start,mid,end):
    temp = []
    i = start
    j = mid+1
    while i<=mid and j<=end:
        if arr[i]<=arr[j]:
            temp.append(arr[i])
            i+=1
        else:
            temp.append(arr[j])
            j+=1
            
    while i<=mid:
        temp.append(arr[i])
        i+=1
        
    while(j<=end):
        temp.append(arr[j])
        j+=1
        
    for idx in range(len(temp)):
        arr[idx+start] = temp[idx]




def MergeSort(arr,start,end):
    if start<end:
        mid = start+(end-start)//2
        MergeSort(arr,start,mid)
        MergeSort(arr,mid+1,end)
        merge(arr,start,mid,end)
        
arr = [5,4,3,2,1]
n = len(arr)
MergeSort(arr,0,n)
# for i in range(len(arr)):
#     print(i)
#     print()